filebeat/input/filestream/copytruncate_prospector.go

```diff
@@ -0,0 +1,344 @@
+					log.Debugf("File %s has been updated", fe.NewPath)
+				}
+
+				if p.fileProspector.isFileIgnored(log, fe, ignoreInactiveSince) {
+					break
+				}
+
+				// check if the event belongs to a rotated file
+				if p.isRotated(fe) {
+					log.Debugf("File %s is rotated", fe.NewPath)
+
+					p.onRotatedFile(log, ctx, fe, src, hg)
+
+				} else {
+					log.Debugf("File %s is original", fe.NewPath)
+					// if file is original, add it to the bookeeper
+					p.rotatedFiles.addOriginalFile(fe.NewPath, src)
+
+					hg.Start(ctx, src)
+				}
+
+			case loginp.OpTruncate:
+				log.Debugf("File %s has been truncated", fe.NewPath)
+
+				s.ResetCursor(src, state{Offset: 0})
+				hg.Restart(ctx, src)
+
+			case loginp.OpDelete:
+				log.Debugf("File %s has been removed", fe.OldPath)
+
+				// if file is rotated, stop harvester and clean up state
+				if p.isRotated(fe) {
+					log.Debugf("Stopping harvester as rotated file %s has been removed.", src.Name())
+
+					hg.Stop(src)
+
+					log.Debugf("Remove state for file as rotated file has been removed: %s", fe.OldPath)
+
+					err := s.Remove(src)
+					if err != nil {
+						log.Errorf("Error while removing state from statestore: %v", err)
+					}
+
+					originalPath := p.rotatedSuffix.ReplaceAllLiteralString(fe.OldPath, "")
+					p.rotatedFiles.removeRotatedFile(originalPath, fe.OldPath)
+				} else {
+					p.fileProspector.onRemove(log, fe, src, s, hg)
+				}
+
+			case loginp.OpRename:
+				log.Debugf("File %s has been renamed to %s", fe.OldPath, fe.NewPath)
+
+				// check if the event belongs to a rotated file
+				if p.isRotated(fe) {
+					log.Debugf("File %s is rotated", fe.NewPath)
+
+					p.onRotatedFile(log, ctx, fe, src, hg)
+				}
+
+				p.fileProspector.onRename(log, ctx, fe, src, s, hg)
+
+			default:
+				log.Error("Unkown return value %v", fe.Op)
+			}
+		}
+		return nil
+	})
+
+	errs := tg.Wait()
+	if len(errs) > 0 {
+		log.Error("%s", sderr.WrapAll(errs, "running prospector failed"))
+	}
+}
+
+func (p *copyTruncateFileProspector) isRotated(event loginp.FSEvent) bool {
+	if p.rotatedSuffix.MatchString(event.NewPath) {
+		return true
+	}
+	return false
+}
+
+func (p *copyTruncateFileProspector) onRotatedFile(
+	log *logp.Logger,
+	ctx input.Context,
+	fe loginp.FSEvent,
+	src loginp.Source,
+	hg loginp.HarvesterGroup,
+) {
+	// Continue reading the rotated file from where we have left off with the original.
+	// The original will be picked up again when updated and read from the beginning.
+	originalPath := p.rotatedSuffix.ReplaceAllLiteralString(fe.NewPath, "")
+	// if we haven't encountered the original file which was rotated, get its information
+	if !p.rotatedFiles.isOriginalAdded(originalPath) {
+		fi, err := os.Stat(originalPath)
+		if err != nil {
+			log.Errorf("Cannot continue file, error while getting the information of the original file: %+v", err)
+			log.Debugf("Starting possibly rotated file from the beginning: %s", fe.NewPath)
+			hg.Start(ctx, src)
```

Comment:
Hm, shall we really try to collect if Stat failed?

Reply:
I choose to collect it to make sure we are not losing any logs until the user notices the error message above. However, I am open to skipping the file, I am just worried about the implications.

Reply:
What are the reasons `err != nil`. Most often I would assume that the file has been removed, or we just don't have permissions. Can you think of others. In general, the harvester would also fail to collect, unless we run into some other issue.

I wonder about network shares. Maybe we just failed because of network errors. Shall we really try to restart from offset 0?

Reply:
In general, removing those copytruncated log files is a mistake. But I see where you are coming from. Then what happens when the file shows up in the end? Shall we scan for rotated files again to try to `Continue` again?

Reply:
Tricky. I would lean towards not collecting and see what happens (if users run into issues).

On the other hand the harvester will fail as well, and if the harvester does not fail we still have the inode check (and in the future maybe some fingerprint?). Given that the harvester should detect if something is odd, we can keep the start here. => Let's keep the code you have written.